parser.c3

```diff
@@ -0,0 +1,658 @@
+}
+
+function parse_scope : (cursor : Cursor&) -> Scope = {
+    runtime_assert(cursor:try_skip_next_and_surrounding_white_space("{"));
+
+    let result : Scope = {make_vector<Line>()};
+    while !cursor:try_skip_next_and_surrounding_white_space("}") {
+        if cursor:get_char() == "\0"[0] {
+            syntax_error(current_char.position, "TODO: unterminated '}'");
+        }
+        result.lines:push_back(cursor:parse_line());
+    }
+    return result;
+}
+
+function parse_graphene_function : (cursor : Cursor&) -> FunctionDefinition = {
+    runtime_assert(cursor:try_skip_next_and_surrounding_white_space("function"));
+
+    // Generic definitions
+    let generics : Vector<GenericDefinition>; // Uninitialized
+
+    if cursor:has_next("[") {
+        generics = cursor:parse_generic_definition_list();
+    } else {
+        // Empty the vector if the function has no generic arguments
+        generics = make_vector<GenericDefinition>(cursor.allocator);
+    }
+
+    cursor:try_skip_white_space();
+
+    // The function's name
+    let name : ProgramSlice = cursor:get_next_ascii_only_identifier({comma: false, operators: false, colon: true});
+    if !cursor:try_skip_next_and_surrounding_white_space(":") {
+        syntax_error(cursor.position, "TODO: expected a ':'");
+    }
+
+    let args : Vector<Argument> = cursor:parse_comma_separated_named_members(make_span(")"));
+
+    if !cursor:try_skip_next_and_surrounding_white_space("->") {
+        syntax_error(cursor.position, "TODO: expected '->'");
+    }
+
+    let return_type : Type = cursor:parse_type();
+    let scope : Scope = cursor:parse_scope();
+    return {0, name, generics, args, return_type, scope};
+}
+
+
+typedef TopLevelToken : {
+    value : ProgramSlice
+}
+
+function parse_next_top_level : (cursor : Cursor&) -> TopLevelToken = {
+    // typedef 'BUG IN SYNAX HIGHLIGHTING... try removing this quoted text'
+    if cursor:is_next("typedef") {
+        // BUG IN SYNTAX HIGHLIGHTING: remove this line
+        let typedef : Typedef& = cursor.allocator:allocate<Typedef>();
+
+        // TODO: this is a copy from stack to heap... is it obvious enough?
+        typedef = cursor:parse_typedef(); // TODO: 'typedef' is highlighted
+        return {2, ref_to_addr(&typedef)};
+    }
+
+    // function
+
+
+    // foreign
+
+    // @implicit
+
+    // @assignment
+
+    // @operator
+
+
+    syntax_error(cursor.position, "Top level features must be one of TODO");
+}
+
+function main : (argc: int, argv: u8[&][&]) -> int = {
+    let file_to_parse : CString = {&argv[1]};
+    let file : File = open(file_to_parse, SYS_O_RDONLY());
+
+
+    let allocator : Allocator = initialize_allocator();
+    let file_contents : Vector<u8> = make_vector<u8>(&allocator);
+
+    // 1) Read the entire file
+    {
+        let read_buffer : u8[1024];
+
+        // TODO: `do while` to avoid temp value? Or `break` to make this prettier
+        let read_data : StringView = make_span("uninitialized");
```

Comment:
something like `"text"sv` would be nice eventually

Reply:
agreed!

Reply:
What do you think of `sv("text")` instead?

It's slightly more verbose but there's no magic because unlike c++ we can actually already pattern match on strings, no need for the extra language feature.